### ISSUE TITLE
Parachain block production stalled after illegal collators joins the network

### ISSUE THREAD

**Issue by stechu** [paritytech/cumulus#541]
The original issue can be found here:
https://github.com/Manta-Network/Manta/issues/127

TLDR: Parachain block production slows down significantly and eventually stalled after illegal collators joins the network.

## Environment
* Polkadot v0.9.8: https://github.com/paritytech/polkadot/releases/tag/v0.9.8
* Cumulus [branch `Polkadot v0.9.8`]: https://github.com/paritytech/cumulus/tree/polkadot-v0.9.8 
* Rust toolchain: `rustc 1.55.0-nightly (7a16cfcff 2021-07-11)`

### The normal path

1. Generate relay chain genesis json file:

```bash
../polkadot/target/release/polkadot build-spec --chain rococo-local --disable-default-bootnode --raw > rococo-local.json
```

2. Run 2 relay-chain validator nodes:

```bash
../polkadot/target/release/polkadot \
--chain rococo-local.json \
--alice \
--tmp \
--ws-port 9941 \
--unsafe-ws-external \
--unsafe-rpc-external \
--rpc-cors=all \
--validator
```
```bash
../polkadot/target/release/polkadot \
--chain rococo-local.json \
--bob \
--tmp \
--ws-port 9942 \
--unsafe-ws-external \
--unsafe-rpc-external \
--rpc-cors=all \
--validator 
```

3. Run one collator node using pre-defined authority account (Alice)

```bash
../cumulus/target/release/polkadot-collator \
--chain=westmint-dev \
--ws-port=9977 \
--port=31201 \
--collator \
--tmp \
--parachain-id=200 \
--alice \
--name=pc1 \
--telemetry-url 'wss://api.telemetry.manta.systems/submit/ 0' \
-- \
--chain=rococo-local.json \
--execution=wasm
```

4. promote Chain 200 to parachain, block production is normal:

```bash
2021-07-20 15:35:06 [Parachain] 🎁 Prepared block for proposing at 58 [hash: 0xe942b1fd20247dce61e07bc7915799d721f19da8be74ddd256e91efb5dba7c32; parent_hash: 0x2737…655c; extrinsics (2): [0x046e…968c, 0x3a04…d873]]
2021-07-20 15:35:06 [Parachain] 🔖 Pre-sealed block for proposal at 58. Hash now 0x8058ecff901f3162868a22a93a4d49be7b2120ac564ead2699e7c78be4bd83de, previously 0xe942b1fd20247dce61e07bc7915799d721f19da8be74ddd256e91efb5dba7c32.
2021-07-20 15:35:06 [Parachain] ✨ Imported #58 (0x8058…83de)
2021-07-20 15:35:06 [Parachain] Produced proof-of-validity candidate. block_hash=0x8058ecff901f3162868a22a93a4d49be7b2120ac564ead2699e7c78be4bd83de
2021-07-20 15:35:06 [Parachain] 💤 Idle (0 peers), best: #57 (0x2737…655c), finalized #55 (0x6ee4…7d23), ⬇ 0 ⬆ 0
2021-07-20 15:35:18 [Parachain] Starting collation. relay_parent=0x98fef7169fce6b3938e9c528843c96479dfc90faec99382ce1340da67b63ffaf at=0x8058ecff901f3162868a22a93a4d49be7b2120ac564ead2699e7c78be4bd83de
2021-07-20 15:35:18 [Parachain] 🙌 Starting consensus session on top of parent 0x8058ecff901f3162868a22a93a4d49be7b2120ac564ead2699e7c78be4bd83de
2021-07-20 15:35:18 [Parachain] Starting collation. relay_parent=0xd8cba663140156df8304f99ebb7b9354b1e5dd7c107cd238770e57436674dd78 at=0x8058ecff901f3162868a22a93a4d49be7b2120ac564ead2699e7c78be4bd83de
2021-07-20 15:35:18 [Parachain] 🎁 Prepared block for proposing at 59 [hash: 0x80e44e641aaac35ea665f574ad017192ae8e24eb7f2913d5ea3f56dad629c11c; parent_hash: 0x8058…83de; extrinsics (2): [0x8f64…633f, 0xf07b…ec2d]]
2021-07-20 15:35:18 [Parachain] 🔖 Pre-sealed block for proposal at 59. Hash now 0xee3b1b8c3eca3fd10c5591cd9e955cdbd8995153683e7072d01a64be3339a748, previously 0x80e44e641aaac35ea665f574ad017192ae8e24eb7f2913d5ea3f56dad629c11c.
2021-07-20 15:35:18 [Parachain] ✨ Imported #59 (0xee3b…a748)
```

### The error path

Repeat 1-2 in the normal path.

3.  Run 3 collator nodes using pre-defined authority account (Alice, Bob, Charlie), 

```bash
../cumulus/target/release/polkadot-collator \
   --chain=westmint-dev \
   --ws-port=9977 \
   --port=31201 \
   --collator \
   --tmp \
   --parachain-id=200 \
   --alice \
   --name=pc1 \
   --telemetry-url 'wss://api.telemetry.manta.systems/submit/ 0' \
   -- \
   --chain=rococo-local.json \
  --execution=wasm
```
```
../cumulus/target/release/polkadot-collator \
  --chain=westmint-dev \
  --ws-port=9988 \
  --port=31202 \
  --collator \
  --tmp \
  --parachain-id=200 \
  --bob \
  --name=pc2 \
  --telemetry-url 'wss://api.telemetry.manta.systems/submit/ 0' \
  -- \
  --chain=rococo-local.json \
  --execution=wasm
```

4. promote Chain 200 to parachain, block production will eventually halt (need to wait a few e.g. 40 - 50 blocks)

```bash
2021-07-20 16:28:02 [Parachain] Starting collation. relay_parent=0xa44ee2619a543550c8bdbf34e7fc835eb6d2ef097a7e3074a16d67d621219ba7 at=0x21fdc1a7423def21f44ca559b2cd651cab4b7502106fcc95f909890df9c926e2
2021-07-20 16:28:02 [Parachain] 🙌 Starting consensus session on top of parent 0x21fdc1a7423def21f44ca559b2cd651cab4b7502106fcc95f909890df9c926e2
2021-07-20 16:28:02 [Parachain] 🎁 Prepared block for proposing at 82 [hash: 0x905c07f5638c1faf657e673478974de1ad89b6df601d1eb727d572460165b7d4; parent_hash: 0x21fd…26e2; extrinsics (2): [0xe378…1d88, 0xc00d…84c1]]
2021-07-20 16:28:02 [Parachain] 🔖 Pre-sealed block for proposal at 82. Hash now 0x2247d73432c94d6f874de9f82fc02c5eb458725c5c81840d2c05782e206d4921, previously 0x905c07f5638c1faf657e673478974de1ad89b6df601d1eb727d572460165b7d4.
2021-07-20 16:28:02 [Parachain] ✨ Imported #82 (0x2247…4921)
2021-07-20 16:28:02 [Parachain] Produced proof-of-validity candidate. block_hash=0x2247d73432c94d6f874de9f82fc02c5eb458725c5c81840d2c05782e206d4921
2021-07-20 16:28:02 [Parachain] 💤 Idle (2 peers), best: #81 (0x21fd…26e2), finalized #80 (0x263a…37f8), ⬇ 0.1kiB/s ⬆ 89 B/s
2021-07-20 16:28:07 [Relaychain] 💤 Idle (2 peers), best: #194 (0xa44e…9ba7), finalized #192 (0x8749…186d), ⬇ 3.1kiB/s ⬆ 3.9kiB/s
2021-07-20 16:28:07 [Parachain] 💤 Idle (2 peers), best: #81 (0x21fd…26e2), finalized #80 (0x263a…37f8), ⬇ 86 B/s ⬆ 1.2kiB/s
2021-07-20 16:28:12 [Relaychain] 💤 Idle (2 peers), best: #194 (0xa44e…9ba7), finalized #192 (0x8749…186d), ⬇ 2.1kiB/s ⬆ 0.3kiB/s
2021-07-20 16:28:12 [Parachain] 💤 Idle (2 peers), best: #81 (0x21fd…26e2), finalized #80 (0x263a…37f8), ⬇ 49 B/s ⬆ 49 B/s
2021-07-20 16:28:17 [Relaychain] 💤 Idle (2 peers), best: #194 (0xa44e…9ba7), finalized #192 (0x8749…186d), ⬇ 2.0kiB/s ⬆ 0.5kiB/s
2021-07-20 16:28:22 [Parachain] 💤 Idle (2 peers), best: #81 (0x21fd…26e2), finalized #80 (0x263a…37f8), ⬇ 32 B/s ⬆ 52 B/s
2021-07-20 16:28:27 [Relaychain] 💤 Idle (2 peers), best: #194 (0xa44e…9ba7), finalized #192 (0x8749…186d), ⬇ 2.0kiB/s ⬆ 0.6kiB/s
2021-07-20 16:28:32 [Relaychain] 💤 Idle (2 peers), best: #194 (0xa44e…9ba7), finalized #192 (0x8749…186d), ⬇ 2.0kiB/s ⬆ 0.3kiB/s
2021-07-20 16:28:37 [Parachain] 💤 Idle (2 peers), best: #81 (0x21fd…26e2), finalized #80 (0x263a…37f8), ⬇ 0 ⬆ 0
2021-07-20 16:28:42 [Relaychain] 💤 Idle (2 peers), best: #194 (0xa44e…9ba7), finalized #192 (0x8749…186d), ⬇ 2.1kiB/s ⬆ 0.3kiB/s
2021-07-20 16:28:42 [Parachain] 💤 Idle (2 peers), best: #81 (0x21fd…26e2), finalized #80 (0x263a…37f8), ⬇ 49 B/s ⬆ 49 B/s
```

<img width="1594" alt="Screen Shot 2021-07-20 at 4 32 00 PM" src="https://user-images.githubusercontent.com/720571/126392273-b624a7d9-8096-4887-b2f4-d366b6aa6056.png">

## Security Implication
This seems give a cheap way to attack single collator parachain  (DoS). This issue is not deterministically reproducible on a more than 2 collator setting.

**Comment by stechu** [comment]
@bkchr Would be great if you can take a look. This may be a cheap way to attack some 1 collator parachains.

**Comment by bkchr** [comment]
Is this still relevant?
